utils/to_string.hh

```diff
@@ -107,45 +90,23 @@ std::ostream& operator<<(std::ostream& os, const std::tuple<T...>& p) {
     return print_tuple(os, p, std::make_index_sequence<sizeof...(T)>());
 }
 
-template <typename T, typename... Args>
-std::ostream& operator<<(std::ostream& os, const std::unordered_set<T, Args...>& items) {
-    os << "{" << utils::join(", ", items) << "}";
-    return os;
-}
-
-template <typename T>
-std::ostream& operator<<(std::ostream& os, const std::set<T>& items) {
-    os << "{" << utils::join(", ", items) << "}";
-    return os;
-}
-
-template<typename T, size_t N>
-std::ostream& operator<<(std::ostream& os, const std::array<T, N>& items) {
-    os << "{" << utils::join(", ", items) << "}";
-    return os;
-}
-
-template <typename K, typename V, typename... Args>
-std::ostream& operator<<(std::ostream& os, const std::unordered_map<K, V, Args...>& items) {
-    os << "{" << utils::join(", ", items) << "}";
+// Exclude string-like types to avoid printing them as vector of chars
+template <std::ranges::range Range>
+requires (
+       !std::convertible_to<Range, std::string>
+    && !std::convertible_to<Range, std::string_view>
```

Comment:
Is sstring "convertible_to" string_view which is why you didn't need to explicitly list it?
If we don't need to explicitly list sstring, maybe we also don't need to explicitly list std::string?